### PR TITLE
Fix signedness and add exceptions 

### PR DESCRIPTION
**Pull request by dschwen**
Replace `int` return codes with exceptions. Keep types sane all the way up to the Fortran calls.

Refs #7041

### CODE REVIEW

**Comment by friedmud** [comment]
This doesn't need any of those `MooseException` specializations.  The only reason to have them is if you were going to do something _different_ with one exception vs another.  As it is... you're just putting a string in... which can be done at construction time of a `MooseException`

File: framework/include/utils/MatrixTools.h [hunk 1]
```diff
@@ -20,15 +21,44 @@ namespace MatrixTools
...
+  MatrixMismatchException() : MooseException("Input and output matrix are not same size square matrices.") {}
+};
+
+class MatrixInversionException : public MooseException
+{
+public:
+  MatrixInversionException(int error) :
+      MooseException(error < 0 ? "Argument " + Moose::stringify(-error) + " was invalid during matrix inversion."
+                               : "Matrix on-diagonal entry " + Moose::stringify(error) + " was exactly zero. Inversion failed.") {}
+};
```
Comment (permcody) [inline framework/include/utils/MatrixTools.h]: I don't see the point in these new exceptions. You aren't catching or handling them so there's little purpose for them over just suppling the string to MooseException. If you actually wanted to catch them.... Well then that would be a different story.

File: unit/src/MatrixToolsTest.C [hunk 2]
```diff
@@ -41,15 +38,13 @@ MatrixToolsTest::matrixInversionTest1()
...
-  CPPUNIT_ASSERT(error == 0);
+  MatrixTools::inverse(mat2, 2);
   CPPUNIT_ASSERT_DOUBLES_EQUAL(1, mat2[0], 1E-5);
   CPPUNIT_ASSERT_DOUBLES_EQUAL(-2, mat2[1], 1E-5);
   CPPUNIT_ASSERT_DOUBLES_EQUAL(-0.5, mat2[2], 1E-5);
   CPPUNIT_ASSERT_DOUBLES_EQUAL(2, mat2[3], 1E-5);
 
-  error = MatrixTools::inverse(m, m);
-  CPPUNIT_ASSERT(error == 0);
+  MatrixTools::inverse(m, m);
```
Comment (friedmud) [inline unit/src/MatrixToolsTest.C]: This is fine as long as you realize that if that `inverse()` throws it's going to cause the current timestep to fail and cutback `dt` and retry.

Is that the right thing to happen?

File: unit/src/MatrixToolsTest.C [hunk 3]
```diff
@@ -79,11 +74,17 @@ MatrixToolsTest::matrixInversionTest2()
   mat3[7] = m[2][1] = 8.0;
   mat3[8] = m[2][2] = 9.0;
 
-  int error = MatrixTools::inverse(mat3, 3);
-  CPPUNIT_ASSERT(error != 0);
+  try
+  {
+    MatrixTools::inverse(mat3, 3);
+  }
+  catch (const MatrixTools::MatrixInversionException & e) {}
```
Comment (friedmud) [inline unit/src/MatrixToolsTest.C]: Do nothing?  So it's ok if it fails?

File: unit/src/MatrixToolsTest.C (hunk 2 above)
Comment (dschwen) [inline unit/src/MatrixToolsTest.C]: This is a unit test.

File: unit/src/MatrixToolsTest.C (hunk 3 above)
Comment (dschwen) [inline unit/src/MatrixToolsTest.C]: This test is not well posed now. The test needs to fail if no exception (or the wrong exception) is thrown.

File: modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C [hunk 4]
```diff
@@ -1375,9 +1375,18 @@ ComputeMultiPlasticityStress::consistentTangentOperator(const RankTwoTensor & st
...
   {
     // invert zzz, in place.  if num_currently_active = 0 then zzz is not needed.
-    const int ierr = MatrixTools::inverse(zzz, num_currently_active);
-    if (ierr != 0)
-      return E_ijkl; // in the very rare case of zzz being singular, just return the "elastic" tangent operator
+    try
+    {
+      MatrixTools::inverse(zzz, num_currently_active);
+    }
+    catch(const MooseException & e)
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C]: This should be a `MatrixTools::MatrixInversionException`. The catch block _explicitly_ deals with singular matrices. Any other exception should make the solve fail IMO.

Comment (friedmud) [inline modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C]: If the exception is for being _singular_ then it should be `MatrixTools::SingularException`

Comment (friedmud) [inline modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C]: Or just `MatrixTools::Singular`

Comment (dschwen) [inline modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C]: Oh. Yeah, sounds good. Unfortunately I just threw out all custom exceptions.

**Comment by WilkAndy** [comment]
That's cool, thanks for doing the int and exception stuff @dschwen ,

a

**Pull request merged**